archunit/src/main/java/com/tngtech/archunit/library/modules/ArchModules.java

```diff
@@ -0,0 +1,556 @@
+import com.google.common.collect.Sets;
+import com.tngtech.archunit.PublicAPI;
+import com.tngtech.archunit.base.ForwardingCollection;
+import com.tngtech.archunit.core.domain.JavaClass;
+import com.tngtech.archunit.core.domain.JavaClasses;
+import com.tngtech.archunit.core.domain.PackageMatcher;
+import com.tngtech.archunit.library.dependencies.Slices;
+import com.tngtech.archunit.library.modules.ArchModule.Identifier;
+import com.tngtech.archunit.library.modules.ArchModules.Creator.WithGenericDescriptor;
+
+import static com.google.common.base.Functions.toStringFunction;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Multimaps.asMap;
+import static com.tngtech.archunit.PublicAPI.Usage.ACCESS;
+import static com.tngtech.archunit.PublicAPI.Usage.INHERITANCE;
+import static com.tngtech.archunit.core.domain.PackageMatcher.TO_GROUPS;
+import static com.tngtech.archunit.library.modules.ArchModule.Identifier.ignore;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A collection of {@link ArchModule "architectural modules"}. This class provides a convenient API to partition the {@link JavaClass classes}
+ * of a code base into (cohesive) modules and assert properties of these modules. E.g. their dependencies to each other or dependencies
+ * not contained in any of the constructed modules.<br>
+ * This class provides several entry points to create {@link ArchModule modules} from a set of {@link JavaClass classes} in a convenient way:<br>
+ * <ul>
+ *     <li>{@link #defineBy(IdentifierAssociation)} - the most generic API allowing a lot of flexibility</li>
+ *     <li>{@link #defineByPackages(String)} - an API similar to {@link Slices#matching(String)}</li>
+ * </ul>
+ */
+@PublicAPI(usage = ACCESS)
+public final class ArchModules<DESCRIPTOR extends ArchModule.Descriptor> extends ForwardingCollection<ArchModule<DESCRIPTOR>> {
+    private final Map<Identifier, ArchModule<DESCRIPTOR>> modulesByIdentifier;
+    private final Map<String, ArchModule<DESCRIPTOR>> modulesByName;
+
+    private ArchModules(Set<ArchModule<DESCRIPTOR>> modules) {
+        this.modulesByIdentifier = modules.stream().collect(toMap(ArchModule::getIdentifier, identity()));
+        this.modulesByName = modules.stream().collect(toMap(ArchModule::getName, identity()));
+
+        SetMultimap<ArchModule.Identifier, ModuleDependency<DESCRIPTOR>> moduleDependenciesByOrigin = HashMultimap.create();
+        modules.forEach(it -> moduleDependenciesByOrigin.putAll(it.getIdentifier(), createModuleDependencies(it, modules)));
+
+        SetMultimap<ArchModule.Identifier, ModuleDependency<DESCRIPTOR>> moduleDependenciesByTarget = HashMultimap.create();
+        moduleDependenciesByOrigin.values().forEach(it -> moduleDependenciesByTarget.put(it.getTarget().getIdentifier(), it));
```

Comment:
I would usually prefer a semantic variable name instead of `it`.
For `modules.forEach(it`, it's maybe easy to see that `it` is a `module`, but I find it more difficult for `moduleDependenciesByOrigin.values()`, so I'd personally go for `moduleDependency` here.

Reply:
For me it depends, if it's very simple I like `it`, but I agree that in this case where we refer to it multiple times this threshold is likely crossed :wink: